az acr login occasionally fails with DOCKER_COMMAND_ERROR in GitHub Actions

Hi @freddydk,

I'm not sure whether this is a bug, but I wanted to report it in case it helps others troubleshoot a similar issue.

I'm getting the following error in the workflow "CreateImages.yml", in the step "Login to ACR":

```
Run az acr login --name $env:ACR_LOGIN_SERVER
WARNING: Registry name is 'fkhcontosoacr'. The following suffix '.azurecr.io' is automatically omitted.
WARNING: You may want to use 'az acr login -n fkhcontosoacr --expose-token' to get a refresh token, which does not require Docker to be installed.
ERROR: 2026-04-30 18:10:50.182405 An error occurred: DOCKER_COMMAND_ERROR
failed to connect to the docker API at npipe:////./pipe/docker_engine; check if the path is correct and if the daemon is running: open //./pipe/docker_engine: The system cannot find the file specified.
Error: Process completed with exit code 1.
```

Just 10 minutes earlier, the same step succeeded in a different run (had the same issue two times now, no root cause to be seen):

```
Run az acr login --name $env:ACR_LOGIN_SERVER
WARNING: Registry name is 'fkhcontosoacr'. The following suffix '.azurecr.io' is automatically omitted.
Login Succeeded
```

--- a/.github/workflows/CreateImages.yml
+++ b/.github/workflows/CreateImages.yml
@@ -137,7 +137,24 @@ jobs:
 
       - name: Login to ACR
         shell: pwsh
-        run: az acr login --name $env:ACR_LOGIN_SERVER
+        run: |
+          $ErrorActionPreference = "Stop"
+          $maxAttempts = 12
+          for ($attempt = 1; $attempt -le $maxAttempts; $attempt++) {
+            docker version *> $null
+            if ($LASTEXITCODE -eq 0) {
+              Write-Host "Docker daemon is ready."
+              az acr login --name $env:ACR_LOGIN_SERVER
+              exit $LASTEXITCODE
+            }
+
+            Write-Host "Docker daemon is not ready yet (attempt $attempt/$maxAttempts). Waiting 10 seconds..."
+            Start-Sleep -Seconds 10
+          }
+
+          Write-Host "Docker daemon did not become ready before ACR login. Last docker version output:"
+          docker version
+          throw "Docker daemon did not become ready before ACR login"
 
       - name: Upload database backup
         shell: pwsh

--- a/.github/workflows/CreateImages.yml
+++ b/.github/workflows/CreateImages.yml
@@ -135,27 +135,6 @@ jobs:
           tenant-id: ${{ env.AZURE_TENANT_ID }}
           subscription-id: ${{ env.AZURE_SUBSCRIPTION_ID }}
 
-      - name: Login to ACR
-        shell: pwsh
-        run: |
-          $ErrorActionPreference = "Stop"
-          $maxAttempts = 12
-          for ($attempt = 1; $attempt -le $maxAttempts; $attempt++) {
-            docker version *> $null
-            if ($LASTEXITCODE -eq 0) {
-              Write-Host "Docker daemon is ready."
-              az acr login --name $env:ACR_LOGIN_SERVER
-              exit $LASTEXITCODE
-            }
-
-            Write-Host "Docker daemon is not ready yet (attempt $attempt/$maxAttempts). Waiting 10 seconds..."
-            Start-Sleep -Seconds 10
-          }
-
-          Write-Host "Docker daemon did not become ready before ACR login. Last docker version output:"
-          docker version
-          throw "Docker daemon did not become ready before ACR login"
-
       - name: Upload database backup
         shell: pwsh
         env:
@@ -240,6 +219,13 @@ jobs:
           $ErrorActionPreference = "Stop"
           Import-Module BcContainerHelper -disableNameChecking
           New-BcImage -artifactUrl $env:ARTIFACT_URL -imageName $env:IMAGE_NAME -baseImage $env:BASE_IMAGE -multitenant:$false
+
+          # Delay ACR login until after the local image exists, so transient
+          # Docker startup issues do not fail the workflow before any registry
+          # access is actually needed.
+          az acr login --name $env:ACR_LOGIN_SERVER
+          if ($LASTEXITCODE -ne 0) { throw "Failed to login to ACR" }
+
           docker tag $env:IMAGE_NAME "$env:ACR_LOGIN_SERVER/businesscentral:$env:IMAGE_TAG"
           docker push "$env:ACR_LOGIN_SERVER/businesscentral:$env:IMAGE_TAG"
           docker tag $env:IMAGE_NAME "$env:ACR_LOGIN_SERVER/businesscentral-$($env:OS_VERSION):$env:IMAGE_TAG"

